src/Compilers/CSharp/Portable/Errors/CustomObsoleteDiagnosticInfo.cs

```diff
@@ -0,0 +1,143 @@
+        {
+            Data = data;
+        }
+
+        private static ErrorCode GetErrorCode(ObsoleteAttributeData data, bool inCollectionInitializer)
+        {
+            // dev11 had a bug in this area (i.e. always produce a warning when there's no message) and we have to match it.
+            if (data.Message is null)
+            {
+                return inCollectionInitializer ? ErrorCode.WRN_DeprecatedCollectionInitAdd : ErrorCode.WRN_DeprecatedSymbol;
+            }
+
+            return (data.IsError, inCollectionInitializer) switch
+            {
+                (true, true) => ErrorCode.ERR_DeprecatedCollectionInitAddStr,
+                (true, false) => ErrorCode.ERR_DeprecatedSymbolStr,
+                (false, true) => ErrorCode.WRN_DeprecatedCollectionInitAddStr,
+                (false, false) => ErrorCode.WRN_DeprecatedSymbolStr
+            };
+        }
+
+        private static object[] GetArguments(Symbol obsoletedSymbol, ObsoleteAttributeData obsoleteAttributeData)
+        {
+            var message = obsoleteAttributeData.Message;
+            if (message is object)
+            {
+                return new object[] { obsoletedSymbol, message };
+            }
+            else
+            {
+                return new object[] { obsoletedSymbol };
+            }
+        }
+
+        public override string MessageIdentifier
+        {
+            get
+            {
+                var id = Data.DiagnosticId;
+                if (!string.IsNullOrEmpty(id))
+                {
+                    return id;
+                }
+
+                return base.MessageIdentifier;
+            }
+        }
+
+        private DiagnosticDescriptor? _descriptor;
+        public override DiagnosticDescriptor Descriptor
+        {
+            get
+            {
+                if (_descriptor == null)
+                {
+                    Interlocked.CompareExchange(ref _descriptor, CreateDescriptor(), null);
+                }
+
+                return _descriptor;
+            }
+        }
+
+        private DiagnosticDescriptor CreateDescriptor()
+        {
+            var baseDescriptor = base.Descriptor;
+            var diagnosticId = Data.DiagnosticId;
+            var urlFormat = Data.UrlFormat;
+            if (diagnosticId is null && urlFormat is null)
+            {
+                return baseDescriptor;
+            }
+
+            var id = MessageIdentifier;
+            var helpLinkUri = baseDescriptor.HelpLinkUri;
+
+            if (urlFormat is object)
+            {
+                try
+                {
+                    helpLinkUri = string.Format(urlFormat, id);
+                }
+                catch
+                {
+                    // TODO: should we report a meta-diagnostic of some kind when the string.Format fails?
+                    // also, should we do some validation of the 'UrlFormat' values provided in source to prevent people from shipping
```

Comment:
Perhaps we could make `ValidateFormatString` aware of this property on ObsoleteAttribute and give a warning if a placeholder greater than 0 is used in the format string.